In [1]:
import keras
keras.__version__

/Users/homme/anaconda/lib/python3.6/site-packages/h5py/__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.
/Users/homme/anaconda/lib/python3.6/importlib/_bootstrap.py:205: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)


'2.2.2'

# Paragraph 1
# 5.1 - Introduction to convnets

This notebook contains the code sample found in Chapter 5, Section 1 of [Deep Learning with Python](https://www.manning.com/books/deep-learning-with-python?a_aid=keras&a_bid=76564dff). Note that the original text features far more content, in particular further explanations and figures: in this notebook, you will only find source code and related comments.

----

First, let's take a practical look at a very simple convnet example. We will use our convnet to classify MNIST digits, a task that you've already been 
through in Chapter 2, using a densely-connected network (our test accuracy then was 97.8%). Even though our convnet will be very basic, its 
accuracy will still blow out of the water that of the densely-connected model from Chapter 2.

The 6 lines of code below show you what a basic convnet looks like. It's a stack of `Conv2D` and `MaxPooling2D` layers. We'll see in a 
minute what they do concretely.
Importantly, a convnet takes as input tensors of shape `(image_height, image_width, image_channels)` (not including the batch dimension). 
In our case, we will configure our convnet to process inputs of size `(28, 28, 1)`, which is the format of MNIST images. We do this via 
passing the argument `input_shape=(28, 28, 1)` to our first layer.

# Paragraph 1
# 5.1 - 컨볼루션 네트워크 소개

이 Notebook은 [Deep Learning with Python](https://www.manning.com/books/deep-learning-with-python?a_aid=keras&a_bid=76564dff)의 5장 1절에 나오는 예제 코드가 포함되어 있습니다.
원본 텍스트(Deep Learning with Python)에 훨씬 더 많은 자료, 특히 추가 설명과 그림들이 포함되어 있습니다.
여기에서는 예제 코드와 코드에 관련된 설명만 제공됩니다.

----

우선, 아주 간단한 컨볼루션 예제를 가지고 실습을 진행해 보겠습니다. 컨볼루션 네트워크를 이용해서 MNIST 숫자를 분류해 볼 것입니다. MNIST 분류 문제는 앞서 2장에서 다뤄본 적이 있습니다. 2장에서는 Densely-connected 네트워크를 이용했습니다(시험 정확도가 97.8%를 기록했습니다.) 이번에 사용할 컨볼루션 네트워크가 아주 단순하기는 해도, 이 네트워크의 정확도는 2장에서 densely-connected 로 구성한 모델의 정확도를 훨씬 뛰어 넘을 수 있을 것입니다. 아래의 6줄 짜리 코드는 기본적인 컨볼루션 네트워크 입니다. `Conv2D`와 `MaxPooling2D` 레이어를 쌓아 올린 모습입니다. 조만간 우리는 이 네트워크가 어떤 일을 해 낼지 구체적으로 살펴봅 것입니다. 중요한 점은, 컨볼루션 네트워크는 입력 텐서의 모양(shape)으로 `(image_heigt, image_width, image_channels)`을 사용합니다(배치의 차원은 고려하지 않았습니다). MNIST를 입력으로 사용하는 우리의 경우에 입력 사이즈는 `(28, 28, 1)` 입니다. 이를 첫 번째 레이어의 인자인 `input_shape = (28, 28, 1)` 을 통해서 명시해줘야 합니다. 

In [3]:
from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

Let's display the architecture of our convnet so far:

앞서 구성한 아키텍쳐를 출력해봅시다.

In [4]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


# Paragraph 2
You can see above that the output of every `Conv2D` and `MaxPooling2D` layer is a 3D tensor of shape `(height, width, channels)`. The width 
and height dimensions tend to shrink as we go deeper in the network. The number of channels is controlled by the first argument passed to 
the `Conv2D` layers (e.g. 32 or 64).

The next step would be to feed our last output tensor (of shape `(3, 3, 64)`) into a densely-connected classifier network like those you are 
already familiar with: a stack of `Dense` layers. These classifiers process vectors, which are 1D, whereas our current output is a 3D tensor. 
So first, we will have to flatten our 3D outputs to 1D, and then add a few `Dense` layers on top:

# Paragraph 2
앞서 우리는 모든 `Conv2D`와 `MaxPooling2D` 레이어가 `(height, width, channels)` 형태의 3D 텐서를 반환함을 확인했습니다. 너비(width)와 높이(height)는 네트워크가 깊어지면 깊어질수록 줄어드는 경향이 있습니다. 채널의 수는 `Conv2D`레이어의 첫 번째 인자로 조절할 수 있습니다(가령 32 또는 64 등으로). 이제 다음 단계는 최종 출력 텐서(`(3, 3, 64)`)를 -우리에게는 이제 익숙한- densely-connected classifier의 입력으로 넣어줄 것입니다 :`Dense` 레이어를 쌓습니다. Densely-connected classifier는 1D 벡터를 가지고 처리하지만, 현재 우리가 가진 출력 값은 3D 텐서입니다.
그렇기 때문에 우선, 3D 텐서를 1D로 펴주는 작업이 필요합니다. 그 이후에 `Dense` 레이어의 입력으로 넣어주도록 합니다. 

In [5]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

We are going to do 10-way classification, so we use a final layer with 10 outputs and a softmax activation. Now here's what our network 
looks like:

우리가 최종적으로 할 것은 10-way classification입니다. 따라서 최종 레이어는 10개의 출력 값과 softmax activation이 필요합니다. 우리의 최종 네트워크는 다음과 같습니다. 

In [6]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 64)                36928     
__________

As you can see, our `(3, 3, 64)` outputs were flattened into vectors of shape `(576,)`, before going through two `Dense` layers.

Now, let's train our convnet on the MNIST digits. We will reuse a lot of the code we have already covered in the MNIST example from Chapter 
2.

`(3, 3, 64)` 출력 텐서가 `Dense` 레이어의 입력으로 들어가기 전 `(576,)` 벡터로 곧게 펴졌음을 볼 수 있습니다. 

자 이제 MNIST 데이터로 컨볼루션 네트워크를 학습시켜 보겠습니다. 2장에서 사용했던 대부분의 코드를 재 사용할 수 있습니다. 

In [7]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [8]:
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
60000/60000 [==============================] - 23s 391us/step - loss: 0.1654 - acc: 0.9480
Epoch 2/5
60000/60000 [==============================] - 23s 385us/step - loss: 0.0454 - acc: 0.9859
Epoch 3/5
60000/60000 [==============================] - 24s 396us/step - loss: 0.0317 - acc: 0.9905
Epoch 4/5
60000/60000 [==============================] - 23s 386us/step - loss: 0.0256 - acc: 0.9926
Epoch 5/5
60000/60000 [==============================] - 24s 400us/step - loss: 0.0197 - acc: 0.9941


Let's evaluate the model on the test data:

시험 데이터로 모델을 평가해봅시다.

In [9]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

10000/10000 [==============================] - 1s 129us/step


In [10]:
test_acc

0.9909

While our densely-connected network from Chapter 2 had a test accuracy of 97.8%, our basic convnet has a test accuracy of 99.3%: we 
decreased our error rate by 68% (relative). Not bad! 

2장의 Densely-connected의 시험 정확도는 97.8% 였습니다. 우리가 이번에 만든 컨볼루션 네트워크의 시험 정확도는 99.3% 입니다: 에러 율이 68% 가량 줄어들었습니다(상대적으로). 나쁘지 않은 결과입니다!